omegaconf/basecontainer.py

```diff
@@ -397,6 +391,17 @@ def _set_item_impl(self, key: Any, value: Any) -> None:
 
         elif isinstance(self.__dict__["_content"], list):
             target_node = isinstance(target_node_ref, ValueNode)
+        should_set_value = target_node_ref is not None and (
+            (
+                # Target node is a container and has an special value or has an
+                # ref_type assigned
+                isinstance(target_node_ref, Container)
+                and (special_value or target_node_ref._has_ref_type())
+            )
+            # if target_node is a valuenode and is not an anynode then
+            # it should set value
```

Comment:
Because AnyNode could be anything from primitive to container?

Reply:
You are not supposed to put containers in AnyNode, containers are ALWAYS converted to DictConfig or ListConfig.

Reply:
Then if you make a config `cfg = OmegaConf.create({"foo": 4})` and `cfg.foo = []` it should wrap a new ListConfig. But if instead the config is:
```
@dataclass
class C:
    foo: Any = 4
cfg= OmegaConf.structured(C)
cfg.foo = []
```
it should give an error?

Reply:
No, it should do the same in both cases.
Why would the second one give an error?

Reply:
Because AnyNode doesn't allow containers.

Reply:
As I said, that container (dict or list) should be wrapped in a DictConfig or a ListConfig and REPLACE the AnyNode.
It can do that because the ref_type allows it.

Reply:
Then an AnyNode will never set its value it will always wrap a new node even if it's a primitive type. ref_type=Any is the default so it should never be a problem wrapping always.